diff --git a/node_modules/@sveltejs/kit/src/runtime/client/fetcher.js b/node_modules/@sveltejs/kit/src/runtime/client/fetcher.js
index 46d7a0b..4ce1955 100644
--- a/node_modules/@sveltejs/kit/src/runtime/client/fetcher.js
+++ b/node_modules/@sveltejs/kit/src/runtime/client/fetcher.js
@@ -160,16 +160,22 @@ function build_selector(resource, opts) {
 	if (opts?.headers || opts?.body) {
 		/** @type {import('types').StrictBody[]} */
 		const values = [];
+    /** @type string | undefined */
+    let hashHeader = undefined;

 		if (opts.headers) {
-			values.push([...new Headers(opts.headers)].join(','));
+      const headers = new Headers(opts.headers);
+
+      hashHeader = headers.get("X-SvelteKit-Hash");
+
+			values.push([...headers].join(','));
 		}

 		if (opts.body && (typeof opts.body === 'string' || ArrayBuffer.isView(opts.body))) {
 			values.push(opts.body);
 		}

-		selector += `[data-hash="${hash(...values)}"]`;
+		selector += `[data-hash="${hashHeader ?? hash(...values)}"]`;
 	}

 	return selector;
diff --git a/node_modules/@sveltejs/kit/src/runtime/server/page/serialize_data.js b/node_modules/@sveltejs/kit/src/runtime/server/page/serialize_data.js
index f879a50..597d4a5 100644
--- a/node_modules/@sveltejs/kit/src/runtime/server/page/serialize_data.js
+++ b/node_modules/@sveltejs/kit/src/runtime/server/page/serialize_data.js
@@ -80,16 +80,22 @@ export function serialize_data(fetched, filter, prerendering = false) {
 	if (fetched.request_headers || fetched.request_body) {
 		/** @type {import('types').StrictBody[]} */
 		const values = [];
+    /** @type string | undefined */
+    let hashHeader = undefined;

 		if (fetched.request_headers) {
-			values.push([...new Headers(fetched.request_headers)].join(','));
+      const headers = new Headers(fetched.request_headers);
+
+      hashHeader = headers.get("X-SvelteKit-Hash");
+
+			values.push([...headers].join(','));
 		}

 		if (fetched.request_body) {
 			values.push(fetched.request_body);
 		}

-		attrs.push(`data-hash="${hash(...values)}"`);
+		attrs.push(`data-hash="${hashHeader ?? hash(...values)}"`);
 	}

 	// Compute the time the response should be cached, taking into account max-age and age.
